### PR TITLE
Don't ask Hive for stats on the partition column

### PR DESCRIPTION
**Pull request by ilfrin**
Filter out the partition that is the partition column from the column
list that is sent in the partition stats request to Hive Metastore,
because this column is not treated as a regular column and is not listed
on the column list, causing some assertions to fail in the JDO code
paths in Hive Metastore (as opposed to the direct SQL code paths). See
Hive Metastore's `ObjectStore.getPartitionColumnStatisticsInternal` for
details.

### CODE REVIEW

File: presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java [hunk 1]
```diff
@@ -54,6 +54,7 @@
 import static java.lang.String.format;
 import static java.util.Objects.requireNonNull;
 import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toSet;
```
Comment (findepi) [inline presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java]: Cmt msg:

> Filter out the partition that is the partition column from the column list ..

"Filter out the columns that are partition columns" ?

File: presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java [hunk 2]
```diff
@@ -100,6 +102,14 @@ public TableStatistics getTableStatistics(ConnectorSession session, ConnectorTab
         return tableStatistics.build();
     }
 
+    private Set<String> filterPartitionColumnFrom(Map<String, ColumnHandle> tableColumns, List<HivePartition> hivePartitions)
```
Comment (findepi) [inline presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java]: put this below usage, i..e below `getPartitionsStatistics` and make static

File: presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java [hunk 3]
```diff
@@ -100,6 +102,14 @@ public TableStatistics getTableStatistics(ConnectorSession session, ConnectorTab
...
+    {
+        return tableColumns.entrySet().stream()
+                .filter(e -> !hivePartitions.get(0).getKeys().containsKey(e.getValue()))
```
Comment (findepi) [inline presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java]: `s/e/entry`

Stupid question. Can `hivePartitions.get(0).getKeys()` be different than `hivePartitions.get(1).getKeys()`?

File: presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java [hunk 4]
```diff
@@ -100,6 +102,14 @@ public TableStatistics getTableStatistics(ConnectorSession session, ConnectorTab
...
+                .map(Map.Entry::getKey)
+                .collect(toSet());
```
Comment (findepi) [inline presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java]: `toImmutableSet`

File: presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java (hunk 2 above)
Comment (losipiuk) [inline presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java]: name it `listNonPartitioningColumns`

File: presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java (hunk 3 above)
Comment (losipiuk) [reply presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java]: Just drop `hivePartitions` arg.
Cast `ColumnHandle` to `HiveColumnHandle`.
And use `HiveColumnHandle.getType()` to determine if column is partitioning column.

File: presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java (hunk 2 above)
Comment (ilfrin) [reply presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java]: done

**Comment by ilfrin** [comment]
AC

File: presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java [hunk 5]
```diff
@@ -73,7 +74,8 @@ public TableStatistics getTableStatistics(ConnectorSession session, ConnectorTab
         if (!isStatisticsEnabled(session)) {
             return TableStatistics.EMPTY_STATISTICS;
         }
-        Map<String, PartitionStatistics> partitionStatistics = getPartitionsStatistics((HiveTableHandle) tableHandle, hivePartitions, tableColumns.keySet());
```
Comment (losipiuk) [inline presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java]: Use plural `partitions` in commit message.

File: presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java [hunk 6]
```diff
@@ -240,13 +242,21 @@ private Estimate summarizePartitionStatistics(
...
-            return ImmutableMap.of(HivePartition.UNPARTITIONED_ID, getTableStatistics(tableHandle.getSchemaTableName(), tableColumns));
+            return ImmutableMap.of(HivePartition.UNPARTITIONED_ID, getTableStatistics(tableHandle.getSchemaTableName(), tableColumns.keySet()));
         }
         else {
-            return getPartitionsStatistics(tableHandle.getSchemaTableName(), hivePartitions, tableColumns);
+            return getPartitionsStatistics(tableHandle.getSchemaTableName(), hivePartitions, listNonPartitioningColumns(tableColumns, hivePartitions));
         }
     }
 
+    private static Set<String> listNonPartitioningColumns(Map<String, ColumnHandle> tableColumns, List<HivePartition> hivePartitions)
```
Comment (losipiuk) [inline presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java]: You were supposed to remote `hivePartitions` arg.

**Comment by ilfrin** [comment]
AC

File: presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java (hunk 6 above)
Comment (ilfrin) [reply presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java]: right, forgot about it

**Comment by findepi** [comment]
Merged, thanks!

**Pull request merged**